interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -0,0 +1,231 @@
+package io.grpc.testing.integration;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.logging.Logger;
+
+public class TrafficControlProxy {
+
+  private static int DEFAULT_BAND = 1024 * 1024; // Bps
+  private static int DEFAULT_DELAY = 100; // ms;
+  private static Logger logger = Logger.getLogger(TrafficControlProxy.class.getName());
+
+  private String loopback = "127.127.127.127";
+  // TODO: make ports arguments
+  private int proxyPort = 5050;
+  private int serverPort = 5001;
+  private int qLength;
+  private int serverQSize;
+  private int clientQSize;
+  private int chunkSize;
+  private boolean shutDown;
+  private LinkedBlockingQueue<Message> serverQueue = new LinkedBlockingQueue<Message>();
+  private LinkedBlockingQueue<Message> clientQueue = new LinkedBlockingQueue<Message>();
+  private ServerSocket client;
+  private Socket server;
+
+  private int bandwidth = DEFAULT_BAND;
+  private int latency = DEFAULT_DELAY;
+
+  public void setBandwidth(int targetBPS) {
+    bandwidth = targetBPS;
+  }
+
+  public void setLatency(int targetMillis) {
+    latency = targetMillis / 2; // divide by two because delay is applied in both directions
+  }
+
+  public void setUp() throws IOException {
+    qLength = Math.max(bandwidth * latency / 1000, 1); // convert from ms to s
+    int maxRead = Math.max(qLength / 2, 1);
+    chunkSize = Math.min(qLength, maxRead);
+
+    logger.info("Started new proxy on port " + proxyPort);
+    client = new ServerSocket(proxyPort);
+    Socket clientSock = client.accept();
+    server = new Socket(loopback, serverPort);
+
+    DataInputStream clientIn = new DataInputStream(clientSock.getInputStream());
+    DataOutputStream clientOut = new DataOutputStream(server.getOutputStream());
+    DataInputStream serverIn = new DataInputStream(server.getInputStream());
+    DataOutputStream serverOut = new DataOutputStream(clientSock.getOutputStream());
+
+    (new Thread(new ClientReader(clientIn))).start();
+    (new Thread(new ClientWriter(clientOut))).start();
+
+    (new Thread(new ServerReader(serverIn))).start();
+    (new Thread(new ServerWriter(serverOut))).start();
+
+  }
+
+  public void shutDown() throws IOException {
+    logger.info("Proxy shutting down... ");
+    shutDown = true;
+    server.close();
+    client.close();
+  }
+
+  private synchronized void incrementServerQSize(int delta) {
+    serverQSize += delta;
+  }
+
+  private synchronized void incrementClientQSize(int delta) {
+    clientQSize += delta;
+  }
+
+  public static void main(String[] args) {
+    TrafficControlProxy p = new TrafficControlProxy();
+    try {
+      p.setUp();
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  private class ClientReader implements Runnable {
```

Comment:
Oh, I see there is `incrementClientQSize` as well. Maybe there should be an additional class that has state for a particular direction and the Runnables?